File: server/BBO/test_functions.py
import numpy as np


def sphere(x: np.ndarray) -> np.ndarray:
    x = np.atleast_2d(x)
    return np.sum(x * x, axis=1)


def rastrigin(x: np.ndarray, A: float = 10.0) -> np.ndarray:
    x = np.atleast_2d(x)
    n = x.shape[1]
    return A * n + np.sum(x * x - A * np.cos(2 * np.pi * x), axis=1)


def ackley(x: np.ndarray, a=20, b=0.2, c=2 * np.pi) -> np.ndarray:
    x = np.atleast_2d(x)
    n = x.shape[1]
    s1 = np.sum(x * x, axis=1) / n
    s2 = np.sum(np.cos(c * x), axis=1) / n
    return -a * np.exp(-b * np.sqrt(s1)) - np.exp(s2) + a + np.e


def bukin6(x: np.ndarray) -> np.ndarray:
    x = np.atleast_2d(x)
    x1 = x[:, 0]
    x2 = x[:, 1]
    return 100.0 * np.sqrt(np.abs(x2 - 0.01 * x1 * x1)) + 0.01 * np.abs(x1 + 10.0)

File: mobile/lib/visualization/test_functions.dart
import 'dart:math' as math;

double sphere(List<double> x) {
  double sum = 0;
  for (final v in x) {
    sum += v * v;
  }
  return sum;
}

double rastrigin(List<double> x, {double a = 10.0}) {
  double sum = a * x.length;
  for (final v in x) {
    sum += v * v - a * math.cos(2 * math.pi * v);
  }
  return sum;
}

double ackley(List<double> x,
    {double a = 20.0, double b = 0.2, double c = 2 * math.pi}) {
  final n = x.length;
  double sumSq = 0;
  double sumCos = 0;
  for (final v in x) {
    sumSq += v * v;
    sumCos += math.cos(c * v);
  }
  return -a * math.exp(-b * math.sqrt(sumSq / n)) -
      math.exp(sumCos / n) +
      a +
      math.e;
}

double bukin6(List<double> x) {
  final x1 = x[0];
  final x2 = x[1];
  return 100 * math.sqrt((x2 - 0.01 * x1 * x1).abs()) +
      0.01 * (x1 + 10).abs();
}

typedef TestFunction = double Function(List<double> x);

TestFunction? getTestFunction(String name) {
  return switch (name) {
    'sphere' => sphere,
    'rastrigin' => rastrigin,
    'ackley' => ackley,
    'bukin6' => bukin6,
    _ => null,
  };
}

const builtinFunctions = ['sphere', 'rastrigin', 'ackley', 'bukin6'];

Map<String, List<double>> defaultBounds = {
  'sphere': [-5.12, 5.12],
  'rastrigin': [-5.12, 5.12],
  'ackley': [-5.0, 5.0],
  'bukin6': [-15.0, -5.0],
};
